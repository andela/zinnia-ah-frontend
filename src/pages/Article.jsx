import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Icon, Item, Modal } from 'semantic-ui-react';
import moment from 'moment';

// components
import Navbar from '../components/presentationals/Navbar/Navbar';
import Loader from '../components/presentationals/Loader/Loader';
import Title from '../components/presentationals/Title/Title';
import Author from '../components/presentationals/Author/Author';
import Button from '../components/presentationals/Button/Button';
import Avatar from '../components/presentationals/Avatar/Avatar';
import CommentSection from '../components/presentationals/Comment/CommentSection';

// modules
import { getSingleArticle } from '../store/modules/article';

// stylesheets
import './Article.scss';

// Images
import { DEFAULT_USER_IMAGE_URL } from '../utils/config';

export class Article extends Component {
  componentDidMount() {
    console.log('Mounted:::');
    const {
      article,
      match: {
        params: { articleId },
      },
      history,
    } = this.props;
    if (!article.id) {
      this.props.getSingleArticle(articleId, history);
    }
  }

  render() {
    const { article, isLoading } = this.props;
    const articleComments = this.props.comments.filter(
      comment => comment.articleId === article.id,
    );
    return (
      <Fragment>
        {isLoading && <Loader size="large" text="loading, please wait" />}
        <Navbar profileUrl={this.props.profileUrl} />
        <div className="article-container" data-test="article">
          {/*sidebar*/}
          <div className="sidebar author">
            <div className="fixed-pos">
              <Author profile={article.author} />
            </div>
          </div>
          {/*content*/}
          <div className="main-content">
            <Item>
              <Title content={article.title} className="title-xl lh-4" />
              <div
                style={{
                  display: 'flex',
                  alignItems: 'center',
                  height: '1.5rem',
                  marginBottom: '1rem',
                }}
              >
                <span>{moment(article.createdAt).fromNow()}</span>
                <span
                  style={{
                    margin: '-.5rem 1rem 0',
                  }}
                >
                  .
                </span>
                <span>{article.readTime} mins read</span>
              </div>
              <div
                style={{
                  display: 'flex',
                  justifyContent: 'start',
                  margingBottom: '2rem',
                }}
              />
              <Item.Content>
                <Item.Image src={article.imageThumbnail} size="massive" />
                <br />
                <Item.Description>
                  <div className="article-body">{article.body}</div>
                </Item.Description>
                <div className="report-rate-container">
                  <div className="rate-inner">
                    <div className="mini-author hide-lg">
                      {article.author && (
                        <div className="d-flex">
                          <Link
                            to={`@${article.author.username}`}
                            className="bg-danger"
                          >
                            <Avatar
                              url={
                                article.author.image || DEFAULT_USER_IMAGE_URL
                              }
                              className="avatar-sm"
                            />
                          </Link>
                          <div>
                            <Link to={`@${article.author.username}`}>
                              <p className="text-center username">
                                @{article.author.username}
                              </p>
                            </Link>
                            <div className="d-flex justify-content-center">
                              <Button
                                className="btn-white btn-sm"
                                value="FOLLOW"
                                type="submit"
                              />
                            </div>
                          </div>
                        </div>
                      )}
                    </div>
                    <div>
                      <Button
                        className="btn-transparent hide-sm"
                        value=""
                        type="button"
                      >
                        <Icon name="flag" color="black" size="big" />
                      </Button>
                    </div>
                  </div>
                  <div className="rate">
                    <p
                      style={{
                        fontSize: '.875rem',
                        lineHeight: '.825rem',
                      }}
                    >
                      Rate this article
                    </p>
                    <div className="d-flex">
                      <Button
                        className="btn-transparent"
                        value=""
                        type="button"
                      >
                        <Icon name="star" color="black" size="big" />
                      </Button>
                      <Button
                        className="btn-transparent"
                        value=""
                        type="button"
                      >
                        <Icon name="star" color="black" size="big" />
                      </Button>
                      <Button
                        className="btn-transparent"
                        value=""
                        type="button"
                      >
                        <Icon name="star" color="black" size="big" />
                      </Button>
                      <Button
                        className="btn-transparent"
                        value=""
                        type="button"
                      >
                        <Icon name="star" color="black" size="big" />
                      </Button>
                      <Button
                        className="btn-transparent"
                        value=""
                        type="button"
                      >
                        <Icon name="star outline" color="black" size="big" />
                      </Button>
                    </div>
                  </div>
                </div>
              </Item.Content>
              <CommentSection
                articleId={article.id || ''}
                comments={articleComments}
              />
            </Item>
          </div>
          {/*right side*/}
          <div className="sidebar actions justify-content-center">
            <div className="fixed-pos" />
          </div>
        </div>

        <div className="footer">
          <div>
            <Button className="btn-transparent" value="" type="button">
              <Icon name="flag" color="black" size="big" />
            </Button>
          </div>
          <div
            className="d-flex"
            style={{
              width: '50%',
            }}
          >
            <div className="mr-auto">
              <Button className="btn-transparent" value="" type="button">
                <Icon name="thumbs up outline" color="black" size="big" />
              </Button>
            </div>
            <div className="mr-auto">
              <Modal
                trigger={
                  <Button className="btn-transparent" type="button" value="">
                    <Icon name="share" color="black" size="big" />
                  </Button>
                }
                closeIcon
                size="tiny"
              >
                <Modal.Header>Share this Article!</Modal.Header>
                <Modal.Content>
                  <div className="share-cta">
                    <a
                      className="links"
                      target="_blank"
                      rel="noopener noreferrer"
                      href={`https://twitter.com/intent/tweet?text=${
                        article.title
                      }&url=${location.href}`}
                    >
                      <Icon name="twitter" size="big" />
                    </a>
                    <a
                      className="links"
                      rel="noopener noreferrer"
                      target="_blank"
                      href={`mailto:''?subject=${article.title}&body=${
                        location.href
                      }`}
                    >
                      <Icon name="mail" color="red" size="big" />
                    </a>
                    <a>
                      <Icon name="facebook" size="big" />
                    </a>
                  </div>
                </Modal.Content>
              </Modal>
            </div>
            <div className="">
              <Button className="btn-transparent" value="" type="button">
                <Icon name="bookmark" color="black" size="big" />
              </Button>
            </div>
          </div>
        </div>
      </Fragment>
    );
  }
}

Article.propTypes = {
  article: PropTypes.object.isRequired,
  getSingleArticle: PropTypes.func,
  isLoading: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      articleId: PropTypes.string.isRequired,
    }),
  }),
  profileUrl: PropTypes.string,
  history: PropTypes.object,
  comments: PropTypes.array.isRequired,
};

Article.defaultProps = {
  profileUrl: '',
};

const mapStateToProps = (state, props) => {
  const { articleId } = props.match.params;
  return {
    isLoading: state.article.isLoading,
    article: state.article.articles[articleId] || {},
    comments: state.comments.comments,
  };
};

export default connect(
  mapStateToProps,
  { getSingleArticle },
)(Article);
