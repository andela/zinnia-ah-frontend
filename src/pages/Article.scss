@import '../assets/variables';

.article-container {
  display: flex;
  padding: 0 6%;

  .sidebar {
    display: flex;
    justify-content: center;
    padding-top: 10.25rem;
  }

  .actions {
    text-align: center;
  }

  .author {
    display: flex;
    justify-content: center;
  }

  .fixed-pos {
    position: fixed;
    //width: 20%;
  }
}

.article-body {
  color: $primary-black;
  font-family: $lato;
  font-size: 1.125rem;
  line-height: 2rem;
  margin-bottom: 6.25rem;
}

.article-cta {
  margin-bottom: 2.5rem;
}

.article-likes-count {
  color: $primary-black;
  font-size: 1.25rem;
  line-height: 1.5rem;
}

.justify-content-center {
  justify-content: center;
}

.main-content {
  display: flex;
  justify-content: center;
  width: 70%;

  .item {
    width: 90%;
  }
}

.share-cta {
  display: flex;
  justify-content: space-around;
}

.report-rate-container {
  border-bottom: .0625rem solid $primary-grey;
  display: flex;
  justify-content: space-between;
  padding-bottom: 1.25rem;
}

.footer {
  background: $primary-white;
  border: 1px solid $primary-black-opacity-1;
  bottom: 0;
  display: none;
  justify-content: space-between;
  padding: 2% 6%;
  position: fixed;
  width: 100%;
}

.mr-auto {
  margin-right: auto;
}

.hide-sm {
  display: block;
}

.hide-lg {
  display: none;
}

.flag:hover {
  color: $primary-red;
}

@media only screen and (max-width: 48rem) {
  .article-container {
    display: block;
    padding: 0 6%;

    .actions {
      display: none;
    }

    .main-content {
      display: flex;
      justify-content: center;
      margin-bottom: 5rem;
      width: 100%;

      .item {
        width: 90%;
      }
    }

    .author {
      display: none;
    }
  }

  .footer {
    display: flex;
  }

  .hide-sm {
    display: none;
  }

  .hide-lg {
    display: block;
  }


}
@media only screen and (max-width: 32rem) {
  .report-rate-container {
    display: block;

    .hide-sm {
      display: block;
    }

    .rate-inner {
      display: flex;
      justify-content: space-between;
    }

    .rate {
      margin-top: 3rem;
    }
  }

  .article-body {
    font-size: 1.125rem;
    line-height: 2rem;
    margin-bottom: 6.25rem;
  }
}
