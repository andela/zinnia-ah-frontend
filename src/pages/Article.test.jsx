import React from 'react';
import { shallow } from 'enzyme';
import checkPropTypes from 'check-prop-types';

import { Article } from './Article';

const props = {
  match: {
    params: {
      articleId: '',
    },
  },
  getSingleArticle: jest.fn(),
  isLoading: false,
  article: {
    likes: [],
  },
};

const loadingProps = {
  match: {
    params: {
      articleId: '',
    },
  },
  getSingleArticle: jest.fn(),
  isLoading: true,
  article: {
    likes: [],
  },
};

describe('Article', () => {
  it('should render without errors', () => {
    const wrapper = shallow(<Article {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render without errors', () => {
    const wrapper = shallow(<Article {...loadingProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive props', () => {
    const expectedProps = { ...props };

    const propsError = checkPropTypes(
      Article.propTypes,
      expectedProps,
      'props',
      Article,
    );

    expect(propsError).toBeUndefined();
  });
});
