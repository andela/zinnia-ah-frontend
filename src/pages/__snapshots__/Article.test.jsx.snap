// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should render without errors 1`] = `
<Fragment>
  <Navbar
    profileUrl=""
  />
  <div
    className="article-container"
    data-test="article"
  >
    <div
      className="sidebar author"
    >
      <div
        className="fixed-pos"
      >
        <Author />
      </div>
    </div>
    <div
      className="main-content"
    >
      <Item>
        <Title
          className="title-xl lh-4"
        />
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "height": "1.5rem",
              "marginBottom": "1rem",
            }
          }
        >
          <span>
            a few seconds ago
          </span>
          <span
            style={
              Object {
                "margin": "-.5rem 1rem 0",
              }
            }
          >
            .
          </span>
          <span>
             mins read
          </span>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "start",
              "margingBottom": "2rem",
            }
          }
        />
        <ItemContent>
          <ItemImage
            size="massive"
          />
          <br />
          <ItemDescription>
            <div
              className="article-body"
            />
          </ItemDescription>
          <div
            className="report-rate-container"
          >
            <div
              className="rate-inner"
            >
              <div
                className="mini-author hide-lg"
              />
              <div>
                <Button
                  className="btn-transparent hide-sm"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="flag"
                    size="big"
                  />
                </Button>
              </div>
            </div>
            <div
              className="rate"
            >
              <p
                style={
                  Object {
                    "fontSize": ".875rem",
                    "lineHeight": ".825rem",
                  }
                }
              >
                Rate this article
              </p>
              <div
                className="d-flex"
              >
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star outline"
                    size="big"
                  />
                </Button>
              </div>
            </div>
          </div>
        </ItemContent>
      </Item>
    </div>
    <div
      className="sidebar actions justify-content-center"
    >
      <div
        className="fixed-pos"
      />
    </div>
  </div>
  <div
    className="footer"
  >
    <div>
      <Button
        className="btn-transparent"
        type="button"
        value=""
      >
        <Icon
          as="i"
          color="black"
          name="flag"
          size="big"
        />
      </Button>
    </div>
    <div
      className="d-flex"
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <div
        className="mr-auto"
      >
        <Button
          className="btn-transparent"
          type="button"
          value=""
        >
          <Icon
            as="i"
            color="black"
            name="thumbs up outline"
            size="big"
          />
        </Button>
      </div>
      <div
        className="mr-auto"
      >
        <Modal
          centered={true}
          closeIcon={true}
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          size="tiny"
          trigger={
            <Button
              className="btn-transparent"
              type="button"
              value=""
            >
              <Icon
                as="i"
                color="black"
                name="share"
                size="big"
              />
            </Button>
          }
        >
          <ModalHeader>
            Share this Article!
          </ModalHeader>
          <ModalContent>
            <div
              className="share-cta"
            >
              <a
                className="links"
                href="https://twitter.com/intent/tweet?text=undefined&url=http://localhost/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <Icon
                  as="i"
                  name="twitter"
                  size="big"
                />
              </a>
              <a
                className="links"
                href="mailto:''?subject=undefined&body=http://localhost/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <Icon
                  as="i"
                  color="red"
                  name="mail"
                  size="big"
                />
              </a>
              <a>
                <Icon
                  as="i"
                  name="facebook"
                  size="big"
                />
              </a>
            </div>
          </ModalContent>
        </Modal>
      </div>
      <div
        className=""
      >
        <Button
          className="btn-transparent"
          type="button"
          value=""
        >
          <Icon
            as="i"
            color="black"
            name="bookmark"
            size="big"
          />
        </Button>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`Article should render without errors 2`] = `
<Fragment>
  <Loader
    size="large"
    text="loading, please wait"
  />
  <Navbar
    profileUrl=""
  />
  <div
    className="article-container"
    data-test="article"
  >
    <div
      className="sidebar author"
    >
      <div
        className="fixed-pos"
      >
        <Author />
      </div>
    </div>
    <div
      className="main-content"
    >
      <Item>
        <Title
          className="title-xl lh-4"
        />
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "height": "1.5rem",
              "marginBottom": "1rem",
            }
          }
        >
          <span>
            a few seconds ago
          </span>
          <span
            style={
              Object {
                "margin": "-.5rem 1rem 0",
              }
            }
          >
            .
          </span>
          <span>
             mins read
          </span>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "start",
              "margingBottom": "2rem",
            }
          }
        />
        <ItemContent>
          <ItemImage
            size="massive"
          />
          <br />
          <ItemDescription>
            <div
              className="article-body"
            />
          </ItemDescription>
          <div
            className="report-rate-container"
          >
            <div
              className="rate-inner"
            >
              <div
                className="mini-author hide-lg"
              />
              <div>
                <Button
                  className="btn-transparent hide-sm"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="flag"
                    size="big"
                  />
                </Button>
              </div>
            </div>
            <div
              className="rate"
            >
              <p
                style={
                  Object {
                    "fontSize": ".875rem",
                    "lineHeight": ".825rem",
                  }
                }
              >
                Rate this article
              </p>
              <div
                className="d-flex"
              >
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star"
                    size="big"
                  />
                </Button>
                <Button
                  className="btn-transparent"
                  type="button"
                  value=""
                >
                  <Icon
                    as="i"
                    color="black"
                    name="star outline"
                    size="big"
                  />
                </Button>
              </div>
            </div>
          </div>
        </ItemContent>
      </Item>
    </div>
    <div
      className="sidebar actions justify-content-center"
    >
      <div
        className="fixed-pos"
      />
    </div>
  </div>
  <div
    className="footer"
  >
    <div>
      <Button
        className="btn-transparent"
        type="button"
        value=""
      >
        <Icon
          as="i"
          color="black"
          name="flag"
          size="big"
        />
      </Button>
    </div>
    <div
      className="d-flex"
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <div
        className="mr-auto"
      >
        <Button
          className="btn-transparent"
          type="button"
          value=""
        >
          <Icon
            as="i"
            color="black"
            name="thumbs up outline"
            size="big"
          />
        </Button>
      </div>
      <div
        className="mr-auto"
      >
        <Modal
          centered={true}
          closeIcon={true}
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          size="tiny"
          trigger={
            <Button
              className="btn-transparent"
              type="button"
              value=""
            >
              <Icon
                as="i"
                color="black"
                name="share"
                size="big"
              />
            </Button>
          }
        >
          <ModalHeader>
            Share this Article!
          </ModalHeader>
          <ModalContent>
            <div
              className="share-cta"
            >
              <a
                className="links"
                href="https://twitter.com/intent/tweet?text=undefined&url=http://localhost/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <Icon
                  as="i"
                  name="twitter"
                  size="big"
                />
              </a>
              <a
                className="links"
                href="mailto:''?subject=undefined&body=http://localhost/"
                rel="noopener noreferrer"
                target="_blank"
              >
                <Icon
                  as="i"
                  color="red"
                  name="mail"
                  size="big"
                />
              </a>
              <a>
                <Icon
                  as="i"
                  name="facebook"
                  size="big"
                />
              </a>
            </div>
          </ModalContent>
        </Modal>
      </div>
      <div
        className=""
      >
        <Button
          className="btn-transparent"
          type="button"
          value=""
        >
          <Icon
            as="i"
            color="black"
            name="bookmark"
            size="big"
          />
        </Button>
      </div>
    </div>
  </div>
</Fragment>
`;
